olive/passes/onnx/peephole_optimizer.py

```diff
@@ -232,13 +235,139 @@ def fuse_reshape_operations(self):
+        raise ValueError(f"No initializer named {name}")
+
+    def _find_value_info_proto_by_name(self, model, name):
+        """Find the ValueInfoProto with the name name in the model's value_info."""
+        for vi in model.graph.value_info:
+            if vi.name == name:
+                return vi
+
+        for initializer in model.graph.initializer:
+            if initializer.name == name:
+                return helper.make_tensor_value_info(name, initializer.data_type, initializer.dims)
+
+        for graph_input in model.graph.input:
+            if graph_input.name == name:
+                return graph_input
+
+        raise ValueError(f"No value info proto named {name}")
+
+    def _run_op(self, model, op):
+        input_names = set()
+
+        op_model = onnx.ModelProto()
+        op_model.ir_version = model.ir_version
+        op_model.producer_name = "constant_folding"
+        op_model.opset_import.extend(model.opset_import)
+        op_model.graph.name = "ConstantFoldingGraph"
+        op_model.graph.node.extend([copy.deepcopy(op)])
+
+        for input_name in op.input:
+            if input_name and input_name not in input_names:
+                try:
+                    initializer = self._find_initializer_by_name(model, input_name)
+                    op_model.graph.initializer.append(copy.deepcopy(initializer))
+                    vi = helper.make_tensor_value_info(initializer.name, initializer.data_type, initializer.dims)
+                    op_model.graph.input.append(vi)
+                except ValueError:
+                    vi = self._find_value_info_proto_by_name(model, input_name)
+                    op_model.graph.input.append(copy.deepcopy(vi))
+                input_names.add(input_name)
+
+        for output_name in op.output:
+            vi = helper.make_tensor_value_info(output_name, TensorProto.UNDEFINED, [])
+            op_model.graph.output.append(vi)
+
+        return op_model
+
+    def _run_onnx_model(self, model):
+        session = ort.InferenceSession(model.SerializeToString(), providers=["CPUExecutionProvider"])
+        input_dict = {}
+        for model_input in session.get_inputs():
+            name = model_input.name
+            tensor = self._find_initializer_by_name(model, name)
+            input_dict[name] = numpy_helper.to_array(tensor)
+        return session.run(None, input_dict)
+
+    def _get_constant_nodes(self, model):
+        dynamic_inputs = {graph_input.name for graph_input in model.graph.input}
+        const_inputs = {
+            initializer.name for initializer in model.graph.initializer if initializer.name not in dynamic_inputs
+        }
+        const_nodes = []
+        for node in model.graph.node:
+            if all(node_input == "" or node_input in const_inputs for node_input in node.input):
+                const_nodes.append(node)
+                const_inputs.update(node.output)
+        return const_nodes
+
+    def fold_constant(self):
```

Comment:
Btw, do you know how fast this optimization step takes? 

I am concerned the iterations through the model proto each time to find initializers, etc might be slow for large models. 
The OnnxDAG I implemented is able to create a dag representation of the model so that we can access the graph components directly.

This could be done later since there is already a TODO for this pass https://github.com/microsoft/Olive/blob/dc6700880d4df013fcfe49f2b6da59c9955ca14f/olive/passes/onnx/peephole_optimizer.py#L23 but it would become more efficient if we used the DAG to find the components we want.

Reply:
I didn't compare these 2, but we can check this when we do this migration. btw I'll remove all TODOs and create tech debt work items for each of them.